src/main/java/frc/robot/subsystems/StateMachine.java

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.commands.States.Ejecting;
+import frc.robot.commands.States.Intaking;
+import frc.robot.commands.States.NoneState;
+import frc.robot.commands.States.PrepShuffle;
+import frc.robot.commands.States.PrepSpeaker;
+import frc.robot.commands.States.StoreFeeder;
+
+public class StateMachine extends SubsystemBase {
+
+  public static RobotState currentState;
+  public static TargetState currentTargetState;
+
+  /** Creates a new StateMachine. */
+  public StateMachine() {
+    currentState = RobotState.NONE;
+    currentTargetState = TargetState.NONE;
+  }
+
+  public void setRobotState(RobotState robotState) {
+    currentState = robotState;
+  }
+
+  public void setTargetState(TargetState targetState) {
+    currentTargetState = targetState;
+  }
+
+  public RobotState getRobotState() {
+    return currentState;
+  }
+
+  public TargetState getTargetState() {
+    return currentTargetState;
+  }
+
+  public Command tryState(RobotState desiredState, StateMachine subStateMachine, Elevator subElevator, Intake subIntake,
+      Transfer subTransfer, Shooter subShooter) {
+    switch (desiredState) {
+      case INTAKING:
+        return new Intaking(subStateMachine, subIntake, subTransfer);
+      // switch (currentState) {
+      // case NONE:
+      // case SHOOTING:
+      // return new Intaking(subStateMachine, subIntake, subTransfer);
+      // }
+
+      case STORE_FEEDER:
+        return new StoreFeeder(subStateMachine, subIntake, subTransfer);
+      // switch (currentState) {
+      // case INTAKING:
+      // return new StoreFeeder(subStateMachine, subIntake, subTransfer);
+      // }
+
+      case EJECTING:
+        return new Ejecting(subStateMachine, subIntake, subTransfer);
+      // switch (currentState) {
+      // case NONE:
+      // case INTAKING:
+      // case STORE_FEEDER:
+      // case PREP_NONE:
+      // case PREP_SHUFFLE:
+      // case PREP_SPEAKER:
+      // return new Ejecting(subStateMachine, subIntake, subTransfer);
+      // }
```

Comment:
So ze problem that I was stuck on for like 3 hours:
I found out that the checks for a valid _currentState_ in the TryState method isn't working (I think because the whole StateMachine subsystem is being initialized once, so _currentState_ is just stuck at NONE and isn't updating).
Commenting out the _currentState_ checks, the state machine works okay (it can go from NONE to INTAKING to STORE_FEEDER to PREP_SHUFFLE/SPEAKER).
With the _currentState_ checks, the state machine goes from NONE to INTAKING to EJECTING (I think because STORE_FEEDER is not valid since currentState is stuck at NONE, so the logic falls all the way down to EJECTING?).
And ye I'm very stuck 🙃

Reply:
Correct assessment of the error. Good job! That is what I determined as well. Adding breaks after the desired state cases will avoid fall through. I have not figured out the solution yet but we could try making things static (static keyword) or try to figure out a way to refresh the try state. Maybe do checks in periodic? Our concept and logic is right but our implantation needs to be rethought.

Reply:
We had a similar (if not this exact) issue in the 2023_Robot_Code with the stowing command of the arm. Back then, our solution was to always return one giant Commands.either that would have to check every state, but I'm not sure how much I enjoy that implementation. Just my 2 cents from what I've seen us do before
Read it for yourself [here](https://github.com/FRCTeam3255/2023_Robot_Code/blob/main/src/main/java/frc/robot/subsystems/Arm.java) (Line 465)

Reply:
> I think because the whole StateMachine subsystem is being initialized once, so _currentState_ is just stuck at NONE and isn't updating
I found the missing ingredient:
```java
gamePieceTrigger
        .onTrue(Commands.deferredProxy(() -> subStateMachine.tryState(RobotState.STORE_FEEDER, subStateMachine,
            subElevator, subIntake, subTransfer,
            subShooter)));
```
The key thing we needed to do was make this a supplier so it was re-evalutated everything instead of once like Alena found.

> I think because STORE_FEEDER is not valid since currentState is stuck at NONE, so the logic falls all the way down to EJECTING?)
Adding breaks after every desiredState case and moving the "default" command outside of the switch statement will fix this